### PR TITLE
State expiry mvp0.1: performance opt, support snapshot prunelevel1

### PR DESCRIPTION
**Pull request by 0xbundler**
### Description

State expiry mvp0.1: performance opt, support snapshot prunelevel1

### Changes

Notable changes: 
* trie/iterator: skip state expiry iter miss node;
* trie/iterator: fix local revive kv wrong issue;
* ethapi: opt getStorageProof in open storage trie;
* trie/tracer: trace node & blob;
* state/stateobject: support snap prune level0&level1;
* prune: opt prune logic for snap;

### CODE REVIEW

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -1169,6 +1170,12 @@ var (
 		Value:    false,
 		Category: flags.StateExpiryCategory,
 	}
+	StateExpiryPruneLevelFlag = &cli.UintFlag{
+		Name:     "state-expiry.prunelevel",
+		Usage:    "set prune level for state expiry,",
```
Comment (asyukii) [inline cmd/utils/flags.go]: small typo, with additional "," at the end.

The correct change would be: "set prune level for state expiry".

File: core/blockchain_reader.go [hunk 2]
```diff
@@ -362,6 +365,24 @@ func (bc *BlockChain) StateAt(startAtRoot common.Hash, startAtBlockHash common.H
...
+	if snap == nil {
+		return nil, fmt.Errorf("cannot open snap to find target storage trie, root: %v, addr: %v", startAtRoot, addr)
+	}
+
+	account, err := snap.Account(crypto.Keccak256Hash(addr[:]))
+	if err != nil {
+		return nil, err
+	}
+
+	return bc.stateCache.OpenStorageTrie(startAtRoot, addr, common.BytesToHash(account.Root))
```
Comment (asyukii) [inline core/blockchain_reader.go:383]: Do we need to add a check if `account.Root` is nil? If it's nil, we will just return `(nil,nil)` or `(nil, error)`.

File: core/state/pruner/pruner.go [hunk 3]
```diff
@@ -65,7 +65,9 @@ const (
 	// to avoid triggering range compaction because of small deletion.
 	rangeCompactionThreshold = 1000000
 
-	FixedPrefixAndAddrSize = 33
+	ContractTrieNodeAvgSize = 105 // This is an experience value
+
+	ContractEpochMetaAvgSize = 16 // This is an experience value
```
Comment (asyukii) [inline core/state/pruner/pruner.go]: Did you mean this is an **estimated** value?

File: core/state/snapshot/snapshot_expire.go [hunk 4]
```diff
@@ -5,31 +5,18 @@ import (
 	"github.com/ethereum/go-ethereum/core/rawdb"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/ethdb"
-	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	ContractSnapshotAvgSize = 17 // This is an experience value
```
Comment (asyukii) [inline core/state/snapshot/snapshot_expire.go]: Same as above

File: core/state/state_expiry.go [hunk 5]
```diff
@@ -28,6 +30,7 @@ type stateExpiryMeta struct {
 	epoch             types.StateEpoch
 	originalRoot      common.Hash
 	originalHash      common.Hash
+	pruneLevel        uint
```
Comment (asyukii) [inline core/state/state_expiry.go]: Would it be better to specify the `uint` type? For example `uint8`

File: core/state/pruner/pruner.go (hunk 3 above)
Comment (0xbundler) [reply core/state/pruner/pruner.go]: yes, it's a estimated value from previous results, it's ok for prune log now.

File: core/blockchain_reader.go (hunk 2 above)
Comment (0xbundler) [reply core/blockchain_reader.go:383]: I think it is better to return an empty trie other than nil, and return err is unnecessary.

**Review by asyukii — APPROVED** [review]

**Pull request merged**